### PR TITLE
Terrain Producer Refactor

### PR DESCRIPTION
**Pull request by ChaoSXDemon**
This branch reworks TerrainProducer such that we can create a consumer on the editor side to create terrain via ECS system.

### CODE REVIEW

File: private/Math/MeshGenerator.cpp [hunk 1]
```diff
@@ -11,7 +11,7 @@
 namespace cd
 {
 
-std::optional<Mesh> MeshGenerator::Generate(const Box& box, const VertexFormat& vertexFormat)
```
Comment (T-rvw) [inline private/Math/MeshGenerator.cpp:14]: We should not use `TBox<float>` directly. Need to revert.

File: private/Math/MeshGenerator.cpp [hunk 2]
```diff
@@ -100,7 +100,7 @@ std::optional<Mesh> MeshGenerator::Generate(const Box& box, const VertexFormat&
 	return mesh;
 }
 
-std::optional<Mesh> MeshGenerator::Generate(const Plane& plane, float width, float height, const VertexFormat& vertexFormat)
```
Comment (T-rvw) [inline private/Math/MeshGenerator.cpp:103]: Ditto. Need to revert.

Comment (T-rvw) [inline private/Math/MeshGenerator.cpp:103]: Ditto. Need to revert.

File: private/Math/MeshGenerator.cpp [hunk 3]
```diff
@@ -140,7 +140,12 @@ std::optional<Mesh> MeshGenerator::Generate(const Plane& plane, float width, flo
...
-std::optional<Mesh> MeshGenerator::Generate(const Sphere& sphere, const VertexFormat& vertexFormat)
+std::optional<Mesh> MeshGenerator::Generate(const TSphere<float>& sphere, const VertexFormat& vertexFormat)
+{
+	return std::nullopt;
+}
+
+std::optional<Mesh> MeshGenerator::GenerateNonUnique(const TBox<float>& box, const VertexFormat& vertexFormat)
 {
 	return std::nullopt;
 }
```
Comment (T-rvw) [inline private/Math/MeshGenerator.cpp]: This function already deleted. Why you need it?

**Comment by T-rvw** [comment]
@ChaoSXDemon As <format> doesn't support in Linux g++ so please use other ways to replace it for `AssetPipelineCore` part.
And MeshGenerator should revert changes.

File: private/Math/MeshGenerator.cpp (hunk 1 above)
Comment (ChaoSXDemon) [reply private/Math/MeshGenerator.cpp:14]: This cause compile error for me. I can try again and paste the error here

File: private/Math/MeshGenerator.cpp (hunk 3 above)
Comment (ChaoSXDemon) [reply private/Math/MeshGenerator.cpp]: Perhaps merge error then, was experiencing linker error

**Comment by ChaoSXDemon** [comment]
@T-rvw , I added implementation of string_format and removed dependency for std::format

**Review by T-rvw — CHANGES_REQUESTED** [review]
Move public/Scene/TerrainTypes.h to TerrainProducer folder? Terrain is a specified feature, not generic to put in the public/scene folder.
And MaterialTextureType I think it doesn't need to add Terrain type. We want to make AssetPipeline look like a very generic lib for scene graph data. Every feature specfied classes and types should put in its Producer/Consumer folder.

Other parts look good to me~

**Comment by ChaoSXDemon** [comment]
I thought we want to edit terrain and that means on the editor side, people need access to the structs in TerrainTypes.h. As for material type, we need that to render terrain as terrain will be constructed with StaticMeshComponent + MaterialComponent and these two alone can't tell the difference between different kinds of meshes.

**Comment by T-rvw** [comment]
> I thought we want to edit terrain and that means on the editor side, people need access to the structs in TerrainTypes.h. As for material type, we need that to render terrain as terrain will be constructed with StaticMeshComponent + MaterialComponent and these two alone can't tell the difference between different kinds of meshes.

Yes. You can also access Terrain type data in the TerrainProducer folder. Just may need to edit editor.lua makefile.

For material type, you may misunderstand something? We will create explicit MaterialType objects in the editor side. See `engine::MaterialType::GetPBRMaterialType()`. You should add a TerrainMaterialType following this pattern. No need to change other things in the AssetPipeline.

After having a special materal type, you can describe feature specified texture type as you like. For example, you can describe heightmap as `MaterialTextureType::AO`. Because terrain material type is only used by terrain renderer, so you can make an implicit rule inside it.

As we will have many rendering features such as Terrain, so we don't want to add many enum values. Instead, features can add their own rules about how to convert MaterialTextureType to the real type they want to use.

**Pull request merged**